pennylane/gradients/parameter_shift.py

```diff
@@ -0,0 +1,450 @@
+    """
+    argnum = argnum or tape.trainable_params
+
+    # Determine the locations of any variance measurements in the measurement queue.
+    var_mask = [m.return_type is qml.operation.Variance for m in tape.measurements]
+    var_idx = np.where(var_mask)[0]
+
+    # Get <A>, the expectation value of the tape with unshifted parameters.
+    expval_tape = tape.copy(copy_operations=True)
+
+    # Convert all variance measurements on the tape into expectation values
+    for i in var_idx:
+        obs = expval_tape._measurements[i].obs
+        expval_tape._measurements[i] = qml.measure.MeasurementProcess(
+            qml.operation.Expectation, obs=obs
+        )
+
+    gradient_tapes = [expval_tape]
+
+    # evaluate the analytic derivative of <A>
+    pdA_tapes, pdA_fn = expval_param_shift(expval_tape, argnum, shift, gradient_recipes, f0)
+    gradient_tapes.extend(pdA_tapes)
+    tape_boundary = len(pdA_tapes) + 1
+
+    # For involutory observables (A^2 = I) we have d<A^2>/dp = 0.
+    # Currently, the only observable we have in PL that may be non-involutory is qml.Hermitian
+    involutory = [i for i in var_idx if tape.observables[i].name != "Hermitian"]
+
+    # If there are non-involutory observables A present, we must compute d<A^2>/dp.
+    non_involutory = set(var_idx) - set(involutory)
+
+    if non_involutory:
+        expval_sq_tape = tape.copy(copy_operations=True)
+
+        for i in non_involutory:
+            # We need to calculate d<A^2>/dp; to do so, we replace the
+            # involutory observables A in the queue with A^2.
+            obs = expval_sq_tape._measurements[i].obs
+            A = obs.matrix
+
+            obs = qml.Hermitian(A @ A, wires=obs.wires)
+            expval_sq_tape._measurements[i] = qml.measure.MeasurementProcess(
+                qml.operation.Expectation, obs=obs
+            )
+
+        # Non-involutory observables are present; the partial derivative of <A^2>
+        # may be non-zero. Here, we calculate the analytic derivatives of the <A^2>
+        # observables.
+        pdA2_tapes, pdA2_fn = expval_param_shift(
+            expval_sq_tape, argnum, shift, gradient_recipes, f0
+        )
+        gradient_tapes.extend(pdA2_tapes)
+
+    def processing_fn(results):
+        mask = qml.math.convert_like(qml.math.reshape(var_mask, [-1, 1]), results[0])
+        f0 = qml.math.expand_dims(results[0], -1)
+
+        pdA = pdA_fn(results[1:tape_boundary])
+        pdA2 = 0
+
+        if non_involutory:
+            # compute the second derivative of non-involutory observables
+            pdA2 = pdA2_fn(results[tape_boundary:])
+
+            if involutory:
+                # if involutory observables are present, ensure they have zero gradient
+                m = [tape.observables[i].name != "Hermitian" for i in var_idx]
+                m = qml.math.convert_like(m, pdA2)
+                pdA2 = qml.math.where(qml.math.reshape(m, [-1, 1]), 0, pdA2)

```

Comment:
Before this logic, what does `pdA2` contain for involutory observables?

Reply:
For the `pdA2_tapes`, we have replaced non-involutory observables with their square (`A` -> `A^2`). However, involutory observables have been left as-is; as a result, they will give the incorrect result! Hence why we need to replace the value with 0 (the known, correct result for involutory variables).

Reply:
Have added a bigger comment here :)